frontend/Obelisk/OAuth/Frontend/Command.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-| Abstract commands needed by obelisk-oauth for its operation.
+
+    The obelisk-oauth frontend is implemented by means of a Requester who is
+    able to execute a simple DSL for sending requests to the backend and
+    storing oauth state to some storage, like session storage.
+
+    Find an example interpreter using session storage in this file.
+-}
+module Obelisk.OAuth.Frontend.Command
+  ( -- * Interpreter implementation
+    --
+    --   you will need to provide an interpreter for the commands in `CommandF`.
+    CommandF (..)
+    -- * Interface
+  , Command
+  , command_storeState
+  , command_loadState
+  , command_removeState
+  , command_getToken
+  ) where
+
+import Control.Monad.Free (Free (..), liftF, MonadFree (..))
```

Comment:
What are the pros/cons of using Free versus asking the user to provide a record of functions?

Reply:
I'll get back to this, have to get the contract sharing going now.

Reply:
@ryantrinkle Sorry for the delay, I really had to get the contract sharing going. I thought using `Requester` would make sense here and needed a data structure in for `Request` that allows me to capture some context at request time and transfer it to response time. I realized that a Free monad (as any other monad) allows me to do that.

Especially with things like `deriving via` I believe a more pleasant API could be provided with a generic monad satisfying some type class like `MonadStore`/`MonadOAuth` or something.

Or of course I could also ditch `Requester` all together and just go with passing in a record of functions and use `performEvent`.

With `deriving via` the `MonadStore` approach could actually be very convenient to use - but probably overkill for this library.

All those approaches are pretty much equivalent, it's a matter of taste and of ergonomics and where we want to go in general with library design. As of now, I would not really go with `Free` as I don't see any real benefits of this approach.

By the way, I came to the conclusion that `Request` should almost always be a monad, at least when used in libraries. Would you agree on that or did I get something wrong?